File: python/src/json_parser.py
from typing import Any, NoReturn

ESCAPES = {'"': '"', "\\": "\\", "/": "/", "b": "\b", "f": "\f", "n": "\n", "r": "\r", "t": "\t"}

def parse_json(source: str) -> Any:
    i = 0

    def fail(msg: str) -> NoReturn:
        raise ValueError(f"{msg} at {i}")

    def peek() -> str:
        return source[i] if i < len(source) else ""

    def eat(c: str) -> bool:
        nonlocal i
        if peek() == c:
            i += 1
            return True
        return False

    def expect(c: str) -> None:
        if not eat(c):
            fail(f"expected {c!r}")

    def ws() -> None:
        nonlocal i
        while peek() in " \t\n\r":
            i += 1

    def literal(word: str, val: Any) -> Any:
        nonlocal i
        if not source.startswith(word, i):
            fail(f"expected {word}")
        i += len(word)
        return val

    def string() -> str:
        nonlocal i
        expect('"')
        out: list[str] = []
        while i < len(source):
            c = source[i]
            i += 1
            if c == '"':
                return "".join(out)
            if c != "\\":
                out.append(c)
                continue
            if i >= len(source) or source[i] not in ESCAPES:
                fail("bad escape")
            out.append(ESCAPES[source[i]])
            i += 1
        fail("unterminated string")

    def number() -> float:
        nonlocal i
        start = i
        eat("-")
        while peek().isdigit():
            i += 1
        if eat("."):
            while peek().isdigit():
                i += 1
        if eat("e") or eat("E"):
            if not eat("+"):
                eat("-")
            while peek().isdigit():
                i += 1
        try:
            return float(source[start:i])
        except ValueError:
            fail("bad number")

    def array() -> list[Any]:
        expect("[")
        items: list[Any] = []
        while True:
            ws()
            if eat("]"):
                return items
            items.append(value())
            ws()
            if not eat(","):
                expect("]")
                return items

    def obj() -> dict[str, Any]:
        expect("{")
        fields: dict[str, Any] = {}
        while True:
            ws()
            if eat("}"):
                return fields
            key = string()
            ws()
            expect(":")
            fields[key] = value()
            ws()
            if not eat(","):
                expect("}")
                return fields

    def value() -> Any:
        ws()
        c = peek()
        if c == "n":
            return literal("null", None)
        if c == "t":
            return literal("true", True)
        if c == "f":
            return literal("false", False)
        if c == '"':
            return string()
        if c == "[":
            return array()
        if c == "{":
            return obj()
        if c == "-" or c.isdigit():
            return number()
        fail("expected value")

    result = value()
    ws()
    if i != len(source):
        fail("trailing input")
    return result


print(parse_json('{"name":"Ada","scores":[1,2,3],"ok":true}'))

File: typescript/src/json-parser.ts
type Json = null | boolean | number | string | Json[] | { [key: string]: Json }

const parseJson = (source: string) => {
  let i = 0

  const peek = () => source[i]
  const eat = (ch: string) => peek() === ch && (i++, true)
  const fail = (message: string): never => {
    throw new SyntaxError(`${message} at ${i}`)
  }
  const ws = () => {
    while (/\s/.test(peek() ?? "")) i++
  }

  const literal = (word: string, value: Json) => {
    if (!source.startsWith(word, i)) fail(`expected ${word}`)
    i += word.length
    return value
  }

  const string = () => {
    if (!eat('"')) fail("expected string")
    let out = ""
    while (i < source.length) {
      const ch = source[i++]
      if (ch === '"') return out
      if (ch !== "\\") {
        out += ch
        continue
      }
      const esc = source[i++]
      out += ({
        '"': '"',
        "\\": "\\",
        "/": "/",
        b: "\b",
        f: "\f",
        n: "\n",
        r: "\r",
        t: "\t"
      } as Record<string, string>)[esc] ?? fail("bad escape")
    }
    throw new SyntaxError(`unterminated string at ${i}`)
  }

  const number = () => {
    const start = i
    while (/[-+.eE\d]/.test(peek() ?? "")) i++
    const value = Number(source.slice(start, i))
    return Number.isFinite(value) ? value : fail("bad number")
  }

  const array = () => {
    eat("[")
    const items: Json[] = []
    while (true) {
      ws()
      if (eat("]")) return items
      items.push(value()!)
      ws()
      if (!eat(",")) {
        if (eat("]")) return items
        fail("expected , or ]")
      }
    }
  }

  const object = () => {
    eat("{")
    const fields: Record<string, Json> = {}
    while (true) {
      ws()
      if (eat("}")) return fields
      const key = string()
      ws()
      if (!eat(":")) fail("expected :")
      fields[key] = value()
      ws()
      if (!eat(",")) {
        if (eat("}")) return fields
        fail("expected , or }")
      }
    }
  }

  const value = () => {
    ws()
    const ch = peek()
    if (ch === "n") return literal("null", null)
    if (ch === "t") return literal("true", true)
    if (ch === "f") return literal("false", false)
    if (ch === '"') return string()
    if (ch === "[") return array()
    if (ch === "{") return object()
    if (ch === "-" || /\d/.test(ch ?? "")) return number()
    throw new SyntaxError(`expected value at ${i}`)
  }

  const result = value()
  ws()
  if (i !== source.length) fail("trailing input")
  return result
}

console.log(parseJson('{"name":"Ada","scores":[1,2,3],"ok":true}'))
